src/transformers/trainer.py

```diff
@@ -3371,6 +3403,9 @@ def training_step(self, model: nn.Module, inputs: Dict[str, Union[torch.Tensor,
             `torch.Tensor`: The tensor with training loss on this batch.
         """
         model.train()
+        if _is_schedule_free_optimizer(self.optimizer):
+            self.optimizer.train()
```

Comment:
Would it be preferable to check if the optimizer has the eval/train methods and call them, instead of checking the class name?

cc @adefazio

Reply:
Yes, the way I do it in my code is by checking for the eval/train methods, i.e. duck typing.

Reply:
is this approach acceptable?

```suggestion
        if hasattr(self.optimizer, 'train') and callable(getattr(self.optimizer, 'train')):
            self.optimizer.train()
```

Reply:
Yes, I think this would be preferable as it's more extensible to other optimizers, and then puts the responsibility for management on implementation / checking on the optimizer implementation, rather than us within trainer